From 859b6bdb8ea67200ce1d5a6295d7f48583528b88 Mon Sep 17 00:00:00 2001
From: Pino Toscano <ptoscano@redhat.com>
Date: Fri, 16 Jan 2015 19:19:16 +0100
Subject: [PATCH 083/118] mllib: add simple qemu filename sanitizing function

It mimics a bit in OCaml the logic used in
commit a95214b1985e694946e3426120a6fdc13a3f081f (for the main library)
and commit 588af1953e5f7ab74009b9175cc5d3efb8bb651a (in disk-create),
so in OCaml tools direct calls to qemu/qemu-img with filenames provided
by the user can properly handle filenames with e.g. ':'.

(cherry picked from commit c796280a8ee45c223a184b956423c55e3af68d81)
---
 mllib/common_utils.ml | 13 +++++++++++++
 1 file changed, 13 insertions(+)

diff --git a/mllib/common_utils.ml b/mllib/common_utils.ml
index 77f2d31..3a33f7f 100644
--- a/mllib/common_utils.ml
+++ b/mllib/common_utils.ml
@@ -419,3 +419,16 @@ let is_char_device file =
 let is_directory path =
   try Sys.is_directory path
   with Sys_error _ -> false
+
+(* Sanitizes a filename for passing it safely to qemu/qemu-img.
+ *
+ * If the filename is something like "file:foo" then qemu-img will
+ * try to interpret that as "foo" in the file:/// protocol.  To
+ * avoid that, if the path is relative prefix it with "./" since
+ * qemu-img won't try to interpret such a path.
+ *)
+let qemu_input_filename filename =
+  if String.length filename > 0 && filename.[0] <> '/' then
+    "./" ^ filename
+  else
+    filename
-- 
2.7.4

